tests/base.py

```diff
@@ -0,0 +1,1034 @@
+        missing_envs = [x for x in ['TAP_SALESFORCE_CLIENT_ID',
+                                    'TAP_SALESFORCE_CLIENT_SECRET',
+                                    'TAP_SALESFORCE_REFRESH_TOKEN']
+                        if os.getenv(x) is None]
+
+        if missing_envs:
+            raise Exception("set environment variables")
+
+    #########################
+    #   Helper Methods      #
+    #########################
+
+    def run_and_verify_check_mode(self, conn_id):
+        """
+        Run the tap in check mode and verify it succeeds.
+        This should be ran prior to field selection and initial sync.
+
+        Return the connection id and found catalogs from menagerie.
+        """
+        # run in check mode
+        check_job_name = runner.run_check_mode(self, conn_id)
+
+        # verify check exit codes
+        exit_status = menagerie.get_exit_status(conn_id, check_job_name)
+        menagerie.verify_check_exit_status(self, exit_status, check_job_name)
+
+        found_catalogs = menagerie.get_catalogs(conn_id)
+        self.assertGreater(len(found_catalogs), 0, msg="unable to locate schemas for connection {}".format(conn_id))
+
+        found_catalog_names = set(map(lambda c: c['tap_stream_id'], found_catalogs))
+
+        self.assertSetEqual(self.expected_streams(), found_catalog_names, msg="discovered schemas do not match")
+        print("discovered schemas are OK")
+
+        return found_catalogs
+
+    def run_and_verify_sync(self, conn_id):
+        """
+        Run a sync job and make sure it exited properly.
+        Return a dictionary with keys of streams synced
+        and values of records synced for each stream
+        """
+        # Run a sync job using orchestrator
+        sync_job_name = runner.run_sync_mode(self, conn_id)
+
+        # Verify tap and target exit codes
+        exit_status = menagerie.get_exit_status(conn_id, sync_job_name)
+        menagerie.verify_sync_exit_status(self, exit_status, sync_job_name)
+
+        # Verify actual rows were synced
+        sync_record_count = runner.examine_target_output_file(
+            self, conn_id, self.expected_streams(), self.expected_primary_keys())
+        self.assertGreater(
+            sum(sync_record_count.values()), 0,
+            msg="failed to replicate any data: {}".format(sync_record_count)
+        )
+        print("total replicated row count: {}".format(sum(sync_record_count.values())))
+
+        return sync_record_count
+
+    @staticmethod
+    def local_to_utc(date: dt):
+        """Convert a datetime with timezone information to utc"""
+        utc = dt(date.year, date.month, date.day, date.hour, date.minute,
+                 date.second, date.microsecond, tz.utc)
+
+        if date.tzinfo and hasattr(date.tzinfo, "_offset"):
+            utc += date.tzinfo._offset
+
+        return utc
+
+    def max_bookmarks_by_stream(self, sync_records):
```

Comment:
If this is unused remove it, but if it's used the for loop inside the for loop might be able to be replaced by a single `max` method.

Reply:
Removed, not used.